src/output.rs

```diff
@@ -353,24 +353,91 @@ impl fmt::Display for DebugBuffer {
     }
 }
 
-fn format_bytes(data: &[u8], f: &mut fmt::Formatter<'_>) -> fmt::Result {
+fn format_bytes(data: &[u8], f: &mut impl fmt::Write) -> fmt::Result {
     #![allow(clippy::assertions_on_constants)]
-    const MIN_OVERFLOW: usize = 8192;
-    const MAX_START: usize = 2048;
-    const MAX_END: usize = 2048;
-    const MAX_PRINTED: usize = MAX_START + MAX_END;
-    assert!(MAX_PRINTED < MIN_OVERFLOW);
-
-    if data.len() >= MIN_OVERFLOW {
-        write!(
-            f,
-            "<{} bytes total>{:?}...<{} bytes omitted>...{:?}",
-            data.len(),
-            data[..MAX_START].as_bstr(),
-            data.len() - MAX_PRINTED,
-            data[data.len() - MAX_END..].as_bstr(),
-        )
+
+    const LINES_MIN_OVERFLOW: usize = 40;
+    const LINES_MAX_START: usize = 10;
+    const LINES_MAX_END: usize = 10;
+    const LINES_MAX_PRINTED: usize = LINES_MAX_START + LINES_MAX_END;
+    assert!(LINES_MAX_PRINTED < LINES_MIN_OVERFLOW);
+
+    const BYTES_MIN_OVERFLOW: usize = 8192;
+    const BYTES_MAX_START: usize = 2048;
+    const BYTES_MAX_END: usize = 2048;
+    const BYTES_MAX_PRINTED: usize = BYTES_MAX_START + BYTES_MAX_END;
+    assert!(BYTES_MAX_PRINTED < BYTES_MIN_OVERFLOW);
+
+    let lines_total = data.as_bstr().lines().count();
+
+    if lines_total >= LINES_MIN_OVERFLOW {
+        let lines_omitted = lines_total - LINES_MAX_PRINTED;
+        let start_lines = data.as_bstr().lines().take(LINES_MAX_START);
+        let end_lines = data.as_bstr().lines().skip(LINES_MAX_START + lines_omitted);
+        write!(f, "<{} lines total>", lines_total)?;
+        write_debug_bstrs(f, start_lines)?;
+        write!(f, "\n<{} lines omitted>\n", lines_omitted)?;
+        write_debug_bstrs(f, end_lines)
+    } else if data.len() >= BYTES_MIN_OVERFLOW {
+        write!(f, "<{} bytes total>", data.len())?;
+        write_debug_bstrs(f, data[..BYTES_MAX_START].lines())?;
+        write!(f, "<{} bytes omitted>", data.len() - BYTES_MAX_PRINTED)?;
+        write_debug_bstrs(f, data[data.len() - BYTES_MAX_END..].lines())
     } else {
-        write!(f, "{:?}", data.as_bstr())
+        write_debug_bstrs(f, data.lines())
+    }
+}
+
+fn write_debug_bstrs<'a>(
+    f: &mut impl std::fmt::Write,
+    lines: impl Iterator<Item = &'a [u8]>,
+) -> std::fmt::Result {
+    let mut newline_needed = false;
+    write!(f, "\"")?;
+    for line in lines {
+        if newline_needed {
+            writeln!(f)?;
+        }
+        let s = format!("{:?}", line.as_bstr());
+        write!(f, "{}", &s[1..s.len() - 1])?;
+        newline_needed = true;
+    }
+    write!(f, "\"")
+}
+
+#[cfg(test)]
+mod test {
+    #[test]
+    fn format_bytes() {
+        let mut s = String::new();
+        for i in 0..40 {
+            s.push_str(&format!("{}\n", i));
+        }
+        let mut buf = String::new();
+        super::format_bytes(s.as_bytes(), &mut buf).unwrap();
+        assert_eq!(
+            r#"<40 lines total>"0
+1
```

Comment:
imo if we keep `<40 lines total>`, it needs to be on its own line

Also, for the quoting, for multiline stuff I think using code fences is more clear.

Reply:
> Also, for the quoting, for multiline stuff I think using code fences is more clear.

By "code fences," you mean ` ``` `, correct?

Would your preference be to use code fences whenever the output spans multiple lines, or only when lines are omitted?